addons/analytic/models/analytic_distribution_model.py

```diff
@@ -59,22 +59,35 @@ def _get_distribution(self, vals):
         res = {}
         for rec in self.search(domain):
             score = 0
-            for key, value in vals.items():
-                if value and rec[key]:
-                    if rec._check_score(key, value) == 1:
-                        score += 1
-                    else:
-                        score = -1
-                        break
+            for key in self._get_fields_to_check():
+                score_to_add = rec._check_score(key, vals.get(key))
+                if score_to_add != -1:
+                    score += score_to_add
+                else:
+                    score = -1
+                    break
```

Comment:
I already said it on the original PR, but I think that this `score = -1` is very C-like, and could be more Pythonic
```diff
diff --git a/addons/analytic/models/analytic_distribution_model.py b/addons/analytic/models/analytic_distribution_model.py
index 25d642b25a7..40e4f22b2de 100644
--- a/addons/analytic/models/analytic_distribution_model.py
+++ b/addons/analytic/models/analytic_distribution_model.py
@@ -4,6 +4,9 @@
 from odoo import api, fields, models, _
 from odoo.exceptions import UserError
 
+class NonMatchingDistribution(Exception):
+    pass
+
 
 class AccountAnalyticDistributionModel(models.Model):
     _name = 'account.analytic.distribution.model'
@@ -57,18 +60,15 @@ class AccountAnalyticDistributionModel(models.Model):
             domain += self._create_domain(fname, value) or []
         best_score = 0
         res = {}
+        fnames = set(self._get_fields_to_check())
         for rec in self.search(domain):
-            score = 0
-            for key in self._get_fields_to_check():
-                score_to_add = rec._check_score(key, vals.get(key))
-                if score_to_add != -1:
-                    score += score_to_add
-                else:
-                    score = -1
-                    break
-            if score > best_score:
-                res = rec.analytic_distribution
-                best_score = score
+            try:
+                score = sum(rec._check_score(key, vals.get(key)) for key in fnames)
+                if score > best_score:
+                    res = rec.analytic_distribution
+                    best_score = score
+            except NonMatchingDistribution:
+                continue
         return res
 
     def _get_fields_to_check(self):
@@ -77,22 +77,19 @@ class AccountAnalyticDistributionModel(models.Model):
     def _check_score(self, key, value):
         self.ensure_one()
         if key == 'company_id':
-            if self[key]:
-                if value == self[key].id:
-                    return 1
-                else:
-                    return -1
-            else:
-                return 0.5
+            if not self.company_id or value == self.company_id.id:
+                return 1 if self.company_id else 0.5
+            raise NonMatchingDistribution
+
         if not self[key]:
             return 0
-        if key == 'partner_category_id':
-            if value and self[key].id in value:
-                return 1
-        if value == self[key].id:
+        if (
+            (value and self[key].id in value)
+            if isinstance(value, (list, tuple)) else
+            (value == self[key].id)
+        ):
             return 1
-        else:
-            return -1
+        raise NonMatchingDistribution
 
     def _create_domain(self, fname, value):
         if not value:
```